### DISCUSSION TITLE
incsearch seems slightly broken?

### DISCUSSION THREAD

**Discussion by pop-cultural-reference** [JetBrains/ideavim#536] [Q&A]
Frequently incsearch seems to be taking me to lines with no match. I think it happens more often at least if the search is at least around 4 characters (probably because the match is then off-screen?). Given a file, search string, and starting point it seems the place it scrolls to is consistent, but eg deleting a character or adding a character sometimes seems to bring me to a real match.

If I press enter though, it always jumps to a real match even if it was on a line with no match.

I've tried commenting out all of my ideavimrc; updating pycharm; uninstalling and reinstalling ideavim; and disabling EasyMotion all without any luck.

Unfortunately I can't share my code, but this is happening often enough that I would have thought other people are running into it?

**Comment by AlexPl292** [comment]
Hi, could you please share your log (https://intellij-support.jetbrains.com/hc/en-us/articles/207241085-Locating-IDE-log-files) ?

**Comment by pop-cultural-reference** [comment]
Definitely: [idea_logs.zip](https://github.com/JetBrains/ideavim/files/243329060104/idea_logs.zip)

Thanks for taking a look!

edit: Ah, sorry for "answering" instead of replying to your answer.

**Reply by AlexPl292** [reply]
Haven't found any suspicious. @citizenmatt Have any ideas regarding that?

**Reply by citizenmatt** [reply]
I haven't seen anything like this, I'm afraid. Just to confirm, typing with `incsearch` will scroll to the wrong place, but hitting enter to finish the search will scroll to the right place?

Could you record your screen showing with a quick example of what's going on, please? Also, is there anything interesting about the setup - is this happening with plain text files, or Python notebooks, do you have custom font settings, are you using an input method editor or searching for non-ASCII characters, etc.?

**Reply by citizenmatt** [reply]
Another possibility - do you have word wrap enabled?

**Reply by pop-cultural-reference** [reply]
No word wrap, no non-ASCII characters I'm aware of, no weird input methods I don't think... Plain .py file (attached a repro.) Font is JetBrains Mono.

The most consistent occurrence of the bug is when I search "test", which highlights a parser.add_argument line without any match in it. Other, correct matches are visible at other lines on the screen, but that's just a coincidence for this particular text file / search text combo (often no match is visible until I hit enter and it has to scroll to a very different spot).

This bug doesn't occur super often, might be specific to particular files.

[buggy_incsearch.py.txt](https://github.com/JetBrains/ideavim/files/9374863454/buggy_incsearch.py.txt)

**Reply by AlexPl292** [reply]
Wow, I really reproduced that with this file. I placed a caret on line 311 and started search. Thank you, I'll take a look!

**Reply by citizenmatt** [reply]
OMG

I have no idea how this bug has not been found before. It's moving the caret/current line to the offset of the first match in the current document in the next open project. 🙀 

I couldn't resist diving in to this one. I couldn't understand what I was seeing in that video - the result on line 328 was being highlighted as the current result, but the current line was being set to 331. These should have been using the same offset. Unless we return a different offset, and that could only happen if we have multiple projects open at the same time. And, well, yeah.

https://user-images.githubusercontent.com/222659/185483858-79131a85-d36f-4dde-a890-5f7211b9ce56.mov

See how the caret line in the editor on the right updates to the same line as the next match found in the editor on the left.

We want IdeaVim to behave like Vim and highlight search results in all open windows (editors) of the document. To get all of the editors, we need an instance of the current project, which we don't have at that point, so we loop over ALL open projects, get the current editor and use that editor's document to find the other editors and highlight everything. But we're missing a few checks, so we a) highlight documents in the wrong project and b) return the next match offset from the wrong editor and therefore c) move the current caret line to the wrong place.

What a brilliant bug 😁

**Reply by AlexPl292** [reply]
Wow, haven't seen Matt that excited :D
The bug is really interesting

**Reply by citizenmatt** [reply]
I do like a good scrolling bug :)

**Reply by citizenmatt** [reply]
I forgot to record an issue for this bug: [VIM-2779](https://youtrack.jetbrains.com/issue/VIM-2779/Incremental-search-can-navigate-to-wrong-offset-with-multiple-projects-open).
